python/IECore/MenuDefinition.py

```diff
@@ -153,22 +153,43 @@ def items( self ) :
 	## Returns a new MenuDefinition containing only the menu items
 	# that reside below the specified root path. The paths in this
 	# new definition are all adjusted to be relative to the requested
-	# root.
+	# root. This also takes into account given `subMenu` definitions,
+	# although the value for subMenu can't be a callable for performance reasons.
 	def reRooted( self, root ) :
 
 		if not root:
 			return MenuDefinition( [] )
 
-		root = root + "/" if not root.endswith("/") else root
+		rootConformed = "/{}/".format(root.strip('/'))
 
 		newItems = []
 		for path, itemDict in self.items() :
-
-			if path.startswith( root ) :
-				newItems.append( ( path[len(root)-1:], itemDict ) )
+			if path.startswith( rootConformed ) :
+				newItems.append( ( path[ len( rootConformed )-1 : ], itemDict ) )
+			elif path == rootConformed[:-1] and isinstance( itemDict.subMenu, MenuDefinition ):
+				# NOTE: subMenu values that are `callable` won't be considered here for performance reasons
```

Comment:
What's the use for this? The main benefit of `subMenu` is that it can be callable, so this seems like a pretty severe restriction. I guess the use case for `getItem()` is to allow someone to modify an existing item, so it wouldn't be useable with callable submenus anyway? Still, this seems like a strange modification to the semantics of the existing `reRooted()` function, just to serve the new `getItem()` function.

This is the one place where I'd have concerns about the impact on Gaffer. We use `reRooted()` in our menus a lot, and it's not clear to me how this will interact. Have you done any interactive testing, with callable and non-callable submenus? Some unit tests might be nice too.

Reply:
The problem is that I would have to resolve definitions and store them so I can actually modify them as well, so it's not ideal too. It's very helpful though that I have the option to retrieve / modify a definition in `subMenu` now, even though I can't do it for those lazy retrieved ones. I tested it with the Jabuka Gaffer UIs and didn't notice any difference in speed. Would it ease your concerns if I made rerooting submenus optional? e.g. `def reRooted( self, root,  resolveSubMenus=False )`

Reply:
My concern wasn't performance, just that it wasn't clear to me whether or not the  change in _behaviour_ here was going to cause problems for Gaffer's use of the MenuDefinition API. Having looked briefly at the code I _think_ it's probably OK, and it'll only change behaviour for malformed definitions where there is an explicit "/foo/bar" path _and_ a "/foo" path with the `.subMenu` field set. But it's late and I might be wrong, so it'd be good to know that you have tested this out in Gaffer yourself, with both sorts of subMenus, and it's all good.

Incidentally, I was going to test it, but GitHub says the PR is coming from "unknown repository" so I don't know how I can pull it.

> Would it ease your concerns if I made rerooting submenus optional? e.g. def reRooted( self, root, resolveSubMenus=False )

Not really. I guess I'm just attached to the simplicity of the old `reRooted()` behaviour, and I think the new behaviour is a bit muddled. The main clients of `reRooted()` are `GafferUI.Menu`, and `IECoreMaya.Menu`, and they don't benefit from this new behaviour at all. If they _did_ want submenus treated this way, then they'd want callables to be taken into account too, because they use `reRooted()` to help translate from the definition to the concrete menu. What I'm trying to say is that `reRooted()` had simple semantics before - just strip off this prefix - and now it has strange semantics.

That's why I was suggesting that `item()` could be implemented without modifying `reRooted()`. For `item()`, ignoring callables seems to be a more reasonable behaviour, because the main clienta (I assume) of `item()` are going to be things that modify items, and it's not possible to modify an item generated by a callable anyway (because it'll be regenerated).

Reply:
Got it. I initially tried to keep code duplication at a minimum, but I agree with you here. I can certainly make it work as part of `item()` and revert `reRooted`. I think my `rootConformed` change in line 163 could stay since it is a bit more robust than the original.